### PR TITLE
Apply reoptimization in PySCIPopt

### PR DESCRIPTION
**Pull request by roessig**
EnableReoptimization added. Needed a fix of the _addLinCons. Other constraints might still not work.
Added a test case for reoptimization with adding of constraints.

The problem was in the
_addLinCons 
function of pyscipopt Model. It used SCIPaddCoefLinear which at some points calls SCIPlockVarCons, where the latter cannot be called in STAGE_PRESOLVED.
I changed _addLinCons not to use SCIPaddCoefLinear anymore but only SCIPcreateConsLinear.

### CODE REVIEW

File: src/pyscipopt/scip.pyx [hunk 1]
```diff
@@ -1153,16 +1156,23 @@ cdef class Model:
         terms = lincons.expr.terms
 
         cdef SCIP_CONS* scip_cons
+
+        cdef int nvars = len(terms.items())
+
+        vars_array = <SCIP_VAR**> malloc(nvars * sizeof(SCIP_VAR*))
+        coeffs_array = <SCIP_Real*> malloc(nvars * sizeof(SCIP_Real))
```
Comment (fserra) [inline src/pyscipopt/scip.pyx]: you have to free the memory

File: src/pyscipopt/scip.pyx [hunk 2]
```diff
@@ -1153,21 +1156,31 @@ cdef class Model:
...
-            self._scip, &scip_cons, str_conversion(kwargs['name']), 0, NULL, NULL,
-            kwargs['lhs'], kwargs['rhs'], kwargs['initial'],
-            kwargs['separate'], kwargs['enforce'], kwargs['check'],
-            kwargs['propagate'], kwargs['local'], kwargs['modifiable'],
-            kwargs['dynamic'], kwargs['removable'], kwargs['stickingatnode']))
+        self._scip, &scip_cons, str_conversion(kwargs['name']), nvars, vars_array, coeffs_array,
+             kwargs['lhs'], kwargs['rhs'], kwargs['initial'],
+             kwargs['separate'], kwargs['enforce'], kwargs['check'],
+             kwargs['propagate'], kwargs['local'], kwargs['modifiable'],
+             kwargs['dynamic'], kwargs['removable'], kwargs['stickingatnode']))
```
Comment (mattmilten) [inline src/pyscipopt/scip.pyx]: This indentation looks strange. Why 5 spaces?

File: tests/test_reopt.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+        m.optimize()
+        print("x", m.getVal(x))
+        print("y", m.getVal(y))
+        self.assertEqual(m.getVal(x), 3.0)
+        self.assertEqual(m.getVal(y), 3.0)
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (mattmilten) [inline tests/test_reopt.py]: There's a newline missing at the end of the file

File: tests/test_reopt.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import unittest
+from pyscipopt import Model
+
+class ReoptimizationTest(unittest.TestCase):
+
+    # function names must start with test
```
Comment (mattmilten) [inline tests/test_reopt.py]: This comment is not necessary.

**Comment by mueldgog** [comment]
I fixed the last remaining things and adapted the code because some names on the master branch have changed.

**Pull request merged**